### PR TITLE
[DOCS] Add Downsampling docs

### PR DESCRIPTION
**Pull request by kilfoyle**
This adds docs for the downsampling project.

**To do:**
 - [x] Add restrictions and other details based on [design doc].(https://github.com/elastic/elasticsearch-adrs/blob/master/analytics/tsdb/tsdb-rollups-design.md).
 - [x] Finish the ILM example after the latest code is merged.
 - [x] Update the code examples. Testing in `master`:
         - ~~the ILM `Rollup` action doesn't seem to work~~
         - ~~`GET /sample-01*/_search` doesn't find the rollup index~~
 - [x] Figure out the code snippet testing.
 - [x] Add Rollup ILM action.
 - [x] Add Downsample API page.
 - [x] Rename a few things (Rollup → Downsample).
 - [x] Mark all docs as Technical Preview.
 - [x] Enable code snippet tests for the Downsample API page.
 - [x] Fix code samples after rollup  → downsample  rename.
 - [x] Store `sample-k8s-metrics.json` on the download site.
 - [x] Open issue/PR for the [suggested follow-on work](https://github.com/elastic/elasticsearch/pull/88571/#issuecomment-1233475660). Opened: https://github.com/elastic/elasticsearch/issues/90065

**Preview**
 - [Downsampling a time series data stream](https://elasticsearch_88571.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/downsampling.html)
 - [Run downsampling manually](https://elasticsearch_88571.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/downsampling-manual.html)
 - [Run downsampling with ILM](https://elasticsearch_88571.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/downsampling-ilm.html)
 - [Downample ILM action](https://elasticsearch_88571.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/ilm-downsample.html)
 - [Downsample API](https://elasticsearch_88571.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/indices-downsample-data-stream.html)

Rel: https://github.com/elastic/elasticsearch/issues/87986

### CODE REVIEW

**Comment by kilfoyle** [comment]
@elasticmachine run elasticsearch-ci/docs

**Comment by kilfoyle** [comment]
@elasticmachine retest this please

**Comment by kilfoyle** [comment]
@elasticmachine update branch

**Comment by debadair** [comment]
The existing Rollup API docs are unrelated to the new downsample née rollup API. Recommend adding a new downsample API ref page and removing the existing Rollup API ref pages from the TOC. While they have yet to be deprecated, we don't want additional folks to start using v1 Rollups. 

As a follow-on, I think we should promote Transforming data to the top level & move [Rolling up historical data](https://www.elastic.co/guide/en/elasticsearch/reference/current/xpack-rollup.html) to the Appendix & exclude it from the TOC. Should also add a cross reference to the new downsampling info for the folks who do find that page.

To exclude a topic from the TOC, add `role="exclude"` to the anchor like we do for the [redirects](https://github.com/elastic/elasticsearch/edit/8.4/docs/reference/redirects.asciidoc).

My understanding is that, at least initially, downsampling downsampled data is not supported. While the downsample action can be used in any phase, there can only be one downsample action in a policy.

**Comment by kilfoyle** [comment]
@elasticmachine update branch

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+`POST /<source-index>/_downsample/<output-downsampled-index>`
+
+[[downsample-api-prereqs]]
+==== {api-prereq-title}
+
+* Only indices in a <<tsds,time series data stream>> are supported.
+
+* If the {es} {security-features} are enabled, you must have the `create_index`
+or `manage` <<privileges-list-indices,index privilege>> for the data stream.
```
Comment (kilfoyle) [inline docs/reference/indices/downsample-data-stream.asciidoc:32]: @csoulios Do you know if I guessed correctly about the required privileges?

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+[[downsampling-ilm-create-index-template]]
+==== Create an index template
+
+This creates an index template for a basic data stream. The available parameters
+for an index template are described in detail in <<set-up-a-data-stream,Set up a
+time series data stream>>.
+
+For simplicity, in the time series mapping all `time_series_metric` parameters
+are set to type `gauge`, but <<time-series-metric,other values>> such as
+`counter` and `histogram` may also be used. The `time_series_metric` values
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling-ilm.asciidoc:90]: `histogram` is not supported. We only support `gauge` and `counter` metric types

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+----
+// TEST[continued]
+
+////
+[source,console]
+----
+DELETE _index_template/*
+----
+// TEST[continued]
+////
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling-ilm.asciidoc:253]: Delete template is not needed.

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+
+[discrete]
+[[downsampling-ilm-ingest-data]]
+==== Ingest time series data
+
+Use a bulk API request automatically create your TSDS and index a set of ten
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling-ilm.asciidoc:259]: I think thi
```suggestion
Use a bulk API request to automatically create your TSDS and index a set of ten
```

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+documents.
+
+**Important:** Before running this bulk request you need to update the
+timestamps to within three to five hours after your current time. That is,
+search `2022-06-21T15` and replace with your present date, and adjust the hour
+to your current time plus three hours.
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling-ilm.asciidoc:265]: Will users have to adjust timestamps? I am afraid this example becomes overly complicated.

File: docs/reference/data-streams/downsampling-manual.asciidoc [hunk 6]
```diff
@@ -0,0 +1,450 @@
+[[downsampling-manual]]
+=== Run downsampling manually
+++++
+<titleabbrev>Run downsampling manually</titleabbrev>
+++++
+
+preview::[]
+
+This is a simplified example that allows you to see quickly how <<downsampling,downsampling>> works to reduce the storage size of a sampled set of metrics. The example uses typical Kubernetes system monitoring data. To test out downsampling, follow these steps:
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling-manual.asciidoc:9]:
```suggestion
This is a simplified example that allows you to see quickly how <<downsampling,downsampling>> works to reduce the storage size of a time series index. The example uses typical Kubernetes cluster monitoring data. To test out downsampling, follow these steps:
```

File: docs/reference/data-streams/downsampling-manual.asciidoc [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+Refer to <<tsds-prereqs,time series data stream prerequisites>>.
+
+For the example you need a sample data file. Download the file from link:
+https://static-www.elastic.co/v3/assets/bltefdd0b53724fa2ce/bltf2fe7a300c3c59f7/631b4bc5cc56115de2f58e8c/sample-k8s-metrics.json[here] and save it in the local directory where you're running {es}.
+
+[discrete]
+[[downsampling-manual-create-index]]
+==== Create a time series index
+
+This creates an index template for a basic data stream. The available parameters for an index template are described in detail in <<set-up-a-data-stream,Set up a time series data stream>>.
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling-manual.asciidoc:30]: This creates the actual index, not an index template

File: docs/reference/data-streams/downsampling-manual.asciidoc [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+[[downsampling-manual-ingest-data]]
+==== Ingest time series data
+
+In a terminal window with {es} running, run the following curl command to load the documents from the downloaded sample data file:
+
+//[source,console]
+//----
+```
+curl -s -H "Content-Type: application/json" \
+   -XPOST http://localhost:9200/sample-01/_bulk?pretty \
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling-manual.asciidoc:201]:
```suggestion
   -XPOST http://<elasticesearch-node>:9200/sample-01/_bulk?pretty \
```

File: docs/reference/data-streams/downsampling-manual.asciidoc [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+
+Downsampling is very easily integrated within an ILM policy. To learn more, try the <<downsampling-ilm,Run downsampling with ILM>> example.
+
+////
+[source,console]
+----
+DELETE /sample-01*
+DELETE _data_stream/*
+DELETE _index_template/*
+----
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling-manual.asciidoc:448]: Not sure this cleaning is required.

File: docs/reference/data-streams/downsampling.asciidoc [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+example, if you run a date_histogram aggregation with `"fixed_interval": "1m"`
+on a downsampled index that has been downsampled at an hourly resolution
+(`"fixed_interval": "1h"`), the query returns one bucket with all of the data at
+minute 0, then 59 empty buckets, and then a bucket with data again for the next
+hour.
+
+NOTE: When runnning queries in {kib} and through Elastic solutions, a normal
+response is returned without notification that some of the queried indices are
+downsampled.
+
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling.asciidoc:126]: We must add the limitations when querying downsample indices:
- Only fixed_interval is supported for date_histogram aggregations
- Only UTC timezone is supported

File: docs/reference/data-streams/downsampling.asciidoc [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+* The new, downsampled index is created on the data tier of the original index
+and it inherits its settings (for example, the number of shards and replicas).
+
+* The numeric `gauge` and `counter` <<mapping-field-meta,metric types>> are
+supported.
+
+* The downsampling configuration is extracted from the time series data stream
+<<tsds-create-mappings-component-template,index mapping>>. The only additional
+required setting is the downsampling `time_interval`.
```
Comment (csoulios) [inline docs/reference/data-streams/downsampling.asciidoc:159]:
```suggestion
required setting is the downsampling `fixed_interval`.
```

**Review by csoulios — COMMENTED** [review]
Thanks for the write up. I have left some comments for fixes and potential improvements.

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 12]
```diff
@@ -0,0 +1,548 @@
...
+[discrete]
+[[downsampling-ilm-policy]]
+==== Create an index lifecycle policy
+
+Create an ILM policy for your time series data. While not required, an ILM
+policy is recommended to automate the management of your time series data stream
+indices.
+
+To enable downsampling, add a <<ilm-downsample,Downsample action>> and set
+<<ilm-downsample-options,`fixed_interval`>> to the frequency at which you want
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc]: Changed to `to the downsampling interval` via [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77)

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+to aggregate the original time series data.
+
+In this example, an ILM policy is configired for the `hot` data tier. A
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc]: Changed to `In this example, an ILM policy is configired for the `hot` phase.` (via [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77))

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+////
+[source,console]
+----
+PUT /_snapshot/found-snapshots
+{
+ "type": "fs",
+  "settings": {
+    "location": "my_backup_location"
+  }
+}
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc]: Nice! I thought this was required for testing, but I've removed it and things are still working. :-)

File: docs/reference/data-streams/downsampling-ilm.asciidoc (hunk 2 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc]: Changed to `For simplicity, in the time series mapping all `time_series_metric` parameters
are set to type `gauge`, but the `counter` metric type may also be used.` (via [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77))

File: docs/reference/data-streams/downsampling-ilm.asciidoc (hunk 4 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc]: Fixed in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77)

File: docs/reference/data-streams/downsampling-manual.asciidoc (hunk 6 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-manual.asciidoc]: Fixed in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77)

File: docs/reference/data-streams/downsampling-manual.asciidoc (hunk 7 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-manual.asciidoc]: Changed to `This creates an index for a basic data stream. The available parameters for an index are described in detail in <<set-up-a-data-stream,Set up a time series data stream>>.` in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77)

File: docs/reference/data-streams/downsampling-manual.asciidoc (hunk 8 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-manual.asciidoc]: Fixed in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77)

File: docs/reference/data-streams/downsampling.asciidoc (hunk 11 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling.asciidoc]: Fixed in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77)

File: docs/reference/ilm/actions/ilm-downsample.asciidoc [hunk 15]
```diff
@@ -0,0 +1,69 @@
+[role="xpack"]
+[[ilm-downsample]]
+=== Downsample
+
+preview::[]
+
+Phases allowed: hot, warm, cold.
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/ilm/actions/ilm-downsample.asciidoc:7]: Added `To use the `downsample` action in the `hot` phase, the `rollover` action *must*
be present. If no rollover action is configured, {ilm-init} will reject the
policy.` via [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77)

File: docs/reference/ilm/actions/ilm-downsample.asciidoc [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+Aggregates a <<tsds,time series data stream>> index and stores
+the results, including statistics about the original data, in a new read-only
+index. For example, you can roll up hourly data into daily or weekly summaries.
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/ilm/actions/ilm-downsample.asciidoc]: Nice! I used this text exactly as is both here and in the Downsample API description.

File: docs/reference/ilm/actions/ilm-downsample.asciidoc [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+This action corresponds to the  <<indices-downsample-data-stream,downsample API>>.
+
+The name of the resulting rollup index is
+`rollup-<original-index-name>-<random-uuid>`. If {ilm-init} performs the
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/ilm/actions/ilm-downsample.asciidoc]: Fixed in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77).

File: docs/reference/ilm/actions/ilm-downsample.asciidoc [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+`rollup` action on a backing index for a data stream, the rollup index is a
+backing index for the same stream.
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/ilm/actions/ilm-downsample.asciidoc]: Fixed in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77).

File: docs/reference/ilm/actions/ilm-downsample.asciidoc [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+
+[role="child_attributes"]
+[[ilm-downsample-options]]
+==== Options
+
+`fixed_interval`:: (Required, string) The
+<<rollup-understanding-group-intervals,fixed time interval>> into which the data
+will be downsampled up.
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/ilm/actions/ilm-downsample.asciidoc]: Fixed in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77).

File: docs/reference/ilm/actions/ilm-downsample.asciidoc [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+{
+  "policy": {
+    "phases": {
+      "delete": {
+        "actions": {
+          "delete" : { }
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/ilm/actions/ilm-downsample.asciidoc]: I have no idea why I added this. :-)
Removed via [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77).

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 21]
```diff
@@ -0,0 +1,64 @@
+[role="xpack"]
+[[indices-downsample-data-stream]]
+=== Downsample API
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc:3]: Added. Good catch!

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+++++
+<titleabbrev>Downsample</titleabbrev>
+++++
+
+Aggregates a <<tsds,time series data stream>> index and stores
+the results, including statistics about the original data, in a new read-only
+index. For example, you can roll up hourly data into daily or weekly summaries.
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc]: Fixed in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77).

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+
+[source,console]
+----
+POST /my-time-series-index/_downsample/my-downsampled-time-series-index
+{
+    "fixed_interval": "1d"
+}
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc]: I've added the `Downsample API` to the `Index Management API` section.

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+* Only indices in a <<tsds,time series data stream>> are supported.
+
+* If the {es} {security-features} are enabled, you must have the `create_index`
+or `manage` <<privileges-list-indices,index privilege>> for the data stream.
+
+[[downsample-api-path-params]]
+==== {api-path-parms-title}
+
+`<source-index>`::
+(Optional, string) Name of the time series index to downsample.
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc:49]: Perfect. I've exactly as suggested.

File: docs/reference/ilm/actions/ilm-downsample.asciidoc [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+
+Phases allowed: hot, warm, cold.
+
+Aggregates a <<tsds,time series data stream>> index and stores
+the results, including statistics about the original data, in a new read-only
+index. For example, you can roll up hourly data into daily or weekly summaries.
+
+This action corresponds to the  <<indices-downsample-data-stream,downsample API>>.
+
+The name of the resulting rollup index is
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/ilm/actions/ilm-downsample.asciidoc]: Ugh! I thought I changed all of these but I missed a whole bunch. It's fixed in [507419d](https://github.com/elastic/elasticsearch/pull/88571/commits/507419dcc6e4498ad50f37505d150e5de4db8a77).

File: docs/reference/data-streams/downsampling-manual.asciidoc (hunk 9 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-manual.asciidoc]: Nice! I thought all the object we created had to be un-created.
Fixed via [347cd81](https://github.com/elastic/elasticsearch/pull/88571/commits/347cd8124f8d37d3efd19095bb154344ed31ce5c)

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 26]
```diff
@@ -0,0 +1,548 @@
...
+[source,console]
+----
+PUT _ilm/policy/datastream_policy   
+{
+  "policy": {                       
+    "phases": {
+      "hot": {                      
+        "actions": {
+          "rollover": {             
+            "max_docs": 1
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc]: Oh, very good point! I updated the policy to use a `max_age` of 5m, which hopefully is a bit more realistic.
```
          "rollover" : {
            "max_age": "5m"
          },
```

File: docs/reference/data-streams/downsampling-ilm.asciidoc (hunk 3 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc:253]: I tried removing this but the CI checks failed, so I'll leave it in unless you have objections.

File: docs/reference/data-streams/downsampling.asciidoc (hunk 10 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling.asciidoc:132]: Added via [9c6c46c](https://github.com/elastic/elasticsearch/pull/88571/commits/9c6c46c2e597177817a1dda6519922f3e54a12c8)

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+
+[role="child_attributes"]
+[[downsample-api-query-parms]]
+==== {api-query-parms-title}
+
+`fixed_interval`:: (Required, <<time-units,time units>>) The interval at which
+to aggregate the original time series data. For example, `60m` produces a
+document for each 60 minute (hourly) interval. This follows standard time
+formatting syntax as used elsewhere in {es}.
++
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc:156]: Great idea. I've updated the Downsampling API page with two short new sections:
 - "The downsampling process"
 - "Source and target index field mappings"

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+[[downsample-api-path-params]]
+==== {api-path-parms-title}
+
+`<source-index>`::
+(Optional, string) Name of the time series index to downsample.
+
+`<output-downsampled_index>`::
++
+--
+(Required, string) Name of the data stream to create. Data stream names must
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc]: Got it. :-)
I've changed this to:

```
<output-downsampled_index>

    (Required, string) Name of the index to create.

    Index names must meet the following criteria:

        Lowercase only
        Cannot include \, /, *, ?, ", <, >, |, ` ` (space character), ,, #
        Indices prior to 7.0 could contain a colon (:), but that’s been deprecated and won’t be supported in 7.0+
        Cannot start with -, _, +
        Cannot be . or ..
        Cannot be longer than 255 bytes (note it is bytes, so multi-byte characters will count towards the 255 limit faster)
        Names starting with . are deprecated, except for [hidden indices](http://localhost:8000/guide/index-modules.html#index-hidden) and internal indices managed by plugins
```

File: docs/reference/data-streams/downsampling-ilm.asciidoc (hunk 5 above)
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc:265]: Yeah, I wish a knew a better approach for this. We have the same problem in the "[Create a TSDS](https://www.elastic.co/guide/en/elasticsearch/reference/master/set-up-tsds.html#create-tsds)" example, where we say `To test the following example, update the timestamps to within the 24 hours after your current time.`

Do you have any ideas? I suppose we could run a script to generate the json request, but that also seems kind of complicated. If you think the need to adjust timestamps makes the example not worth including, I can remove it.

**Review by csoulios — COMMENTED** [review]
@kilfoyle thanks for all the changes. I like how it all comes together.
I have left a few more comments. After those, I think it should be ready

File: docs/reference/data-streams/downsampling.asciidoc [hunk 29]
```diff
@@ -0,0 +1,177 @@
...
+You can use the <<search-search,`_search`>> and <<async-search,`_async_search`>>
+endpoints to query a downsampled index. Multiple raw data and downsampled
+indices can be queried in a single request, and a single request can include
+downsampled indices at different granularities (different bucket timespan). That
+is, you can query data streams that contain downsampled indices with multiple
+downsampling intervals (for example, `15m`, `1h`, `1d`).
+
+The result of a time based histogram aggregation is in a uniform bucket size and
+each downsampled index returns data ignoring the downsampling time interval. For
+example, if you run a date_histogram aggregation with `"fixed_interval": "1m"`
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling.asciidoc]: Fixed in [deefdcd](https://github.com/elastic/elasticsearch/pull/88571/commits/deefdcdb9c57a0a0d1c52d694c3b410194e7e884).

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 30]
```diff
@@ -0,0 +1,548 @@
...
+    "phases": {
+      "hot": {
+        "actions": {
+          "rollover" : {
+            "max_age": "5m"
+          },
+          "downsample": {
+  	          "fixed_interval": "1h"
+  	      }
+        }
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc]: Nice catch!
Fixed in [deefdcd](https://github.com/elastic/elasticsearch/pull/88571/commits/deefdcdb9c57a0a0d1c52d694c3b410194e7e884).

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 31]
```diff
@@ -0,0 +1,548 @@
...
+[source,console]
+----
+PUT _cluster/settings
+{
+  "transient": {
+    "indices.lifecycle.poll_interval": "1m"
+  }
+}
+----
+// TEST[continued]
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc]: Good point! I removed this section about updating the poll interval, and instead recommended users to wait ten minutes after the documents are loaded before querying for the updated results.

> By default, index lifecycle management checks every ten minutes for indices that
> meet policy criteria. Wait for about ten minutes (maybe brew up a quick coffee
> or tea &#9749; ) and then re-run the `GET _data_stream` request.

Fixed in [deefdcd](https://github.com/elastic/elasticsearch/pull/88571/commits/deefdcdb9c57a0a0d1c52d694c3b410194e7e884).

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 32]
```diff
@@ -0,0 +1,157 @@
+[role="xpack"]
+[[indices-downsample-data-stream]]
+=== Downsample API
+++++
+<titleabbrev>Downsample</titleabbrev>
+++++
+
+preview::[]
+
+Aggregates a <<tsds,time series data stream>> (TSDS) index and stores
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc]: 👍  Changed to `Aggregates a time series (TSDS) index and stores
pre-computed statistical summaries (`min`, `max`, `sum`, `value_count` and
`avg`) for each metric field grouped by a configured time interval.`

Fixed in [deefdcd](https://github.com/elastic/elasticsearch/pull/88571/commits/deefdcdb9c57a0a0d1c52d694c3b410194e7e884).

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 33]
```diff
@@ -0,0 +1,157 @@
...
+
+[[downsample-api-request]]
+==== {api-request-title}
+
+`POST /<source-index>/_downsample/<output-downsampled-index>`
+
+[[downsample-api-prereqs]]
+==== {api-prereq-title}
+
+* Only indices in a <<tsds,time series data stream>> are supported.
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc:86]: 👍  Changed to `Aggregates a time series (TSDS) index and stores
pre-computed statistical summaries (`min`, `max`, `sum`, `value_count` and
`avg`) for each metric field grouped by a configured time interval.`

Fixed in [deefdcd](https://github.com/elastic/elasticsearch/pull/88571/commits/deefdcdb9c57a0a0d1c52d694c3b410194e7e884).

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 34]
```diff
@@ -0,0 +1,157 @@
+[role="xpack"]
+[[indices-downsample-data-stream]]
+=== Downsample API
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc]: Fixed in [deefdcd](https://github.com/elastic/elasticsearch/pull/88571/commits/deefdcdb9c57a0a0d1c52d694c3b410194e7e884).

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 35]
```diff
@@ -0,0 +1,157 @@
...
+++++
+
+preview::[]
+
+Aggregates a <<tsds,time series data stream>> (TSDS) index and stores
+pre-computed statistical summaries (`min`, `max`, `sum`, `value_count` and
+`avg`) for each metric field grouped by a configured time interval. For example,
+a TSDS index that contains metrics sampled every 10 seconds can be downsampled
+to an hourly index. All documents within an hour interval are summarized and
+stored as a single document and stored in the downsample index.
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc]: Fixed in [deefdcd](https://github.com/elastic/elasticsearch/pull/88571/commits/deefdcdb9c57a0a0d1c52d694c3b410194e7e884).

File: docs/reference/data-streams/downsampling-ilm.asciidoc [hunk 36]
```diff
@@ -0,0 +1,548 @@
...
+++++
+<titleabbrev>Run downsampling with ILM</titleabbrev>
+++++
+
+preview::[]
+
+This is a simplified example that allows you to see quickly how
+<<downsampling,downsampling>> works as part of an ILM policy to reduce the
+storage size of a sampled set of metrics. The example uses typical Kubernetes
+system monitoring data. To test out downsampling with ILM, follow these steps:
```
[the remark this answers is not in the document]
Comment (kilfoyle) [reply docs/reference/data-streams/downsampling-ilm.asciidoc]: Fixed in [deefdcd](https://github.com/elastic/elasticsearch/pull/88571/commits/deefdcdb9c57a0a0d1c52d694c3b410194e7e884).

**Comment by kilfoyle** [comment]
@csoulios All fixed! I'm really glad that things look okay. :-)

File: docs/reference/data-streams/downsampling-ilm.asciidoc (hunk 5 above)
Comment (csoulios) [reply docs/reference/data-streams/downsampling-ilm.asciidoc:250]: Unfortunately, I don't have anything better. If it works for the "Create TSDS" example, let's keep it here too.

File: docs/reference/indices/downsample-data-stream.asciidoc [hunk 37]
```diff
@@ -0,0 +1,157 @@
...
+[[downsample-api-request]]
+==== {api-request-title}
+
+`POST /<source-index>/_downsample/<output-downsampled-index>`
+
+[[downsample-api-prereqs]]
+==== {api-prereq-title}
+
+* Only indices in a <<tsds,time series data stream>> are supported.
+
```
Comment (salvatore-campagna) [inline docs/reference/indices/downsample-data-stream.asciidoc:87]: When it comes to permissions requirements are the following:
* `all` or `manage` are required (those for instance, as satisfied by an admin user normally)
* no Field Level Security and no Document Level Secuirty should be defined on the source index. We check if FLS or DLS is defined on the source idnex and we fail downsampling if that is the case.

Comment (csoulios) [reply docs/reference/indices/downsample-data-stream.asciidoc:87]: Thanks for chiming in, Salvatore

**Review by csoulios — APPROVED** [review]
LGTM! 

There's one last fix for the required permissions (please see Salvatore's comment). 
I don't think it needs another review after that.

Thanks for all the iterations, David

File: docs/reference/indices/downsample-data-stream.asciidoc (hunk 37 above)
Comment (kilfoyle) [reply docs/reference/indices/downsample-data-stream.asciidoc:87]: Thanks Salvatore!
The prereqs are now:

* Only indices in a <<tsds,time series data stream>> are supported.

* If the {es} {security-features} are enabled, you must have the `all`
or `manage` <<privileges-list-indices,index privilege>> for the data stream.

* Neither <<field-and-document-access-control,field nor document level security>> can be defined on the source index.

* The source index must be read only (`index.blocks.write: true`).

**Pull request merged**